### PR TITLE
Github App Authentication

### PR DESCRIPTION
**Pull request by saisankargochhayat**
## Related Issues and Dependencies
Related - https://github.com/thoth-station/kebechet/issues/585

This PR adds source-management to be able to create PR's as a Github Application
Added refresh token logic

### CODE REVIEW

**Comment by saisankargochhayat** [comment]
@fridex Can you please review this

File: thoth/sourcemanagement/github_authentication.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+_BASE_URL = "https://api.github.com/"
+
+
+class GithubAuthentication:
+    """Handles getting the OAuth Token for the Github Application."""
+
+    GITHUB_PRIVATE_KEY_PATH = str(os.getenv("GITHUB_PRIVATE_KEY_PATH"))
+    _FILE_PATH = Path(GITHUB_PRIVATE_KEY_PATH)
+    GITHUB_PRIVATE_KEY = open(_FILE_PATH, "r").read()
```
Comment (fridex) [inline thoth/sourcemanagement/github_authentication.py]: Please do this in a method. The file description will stay open, moreover, this code is executed on the import every time which can have negative performance side effects (please do it lazily).

File: thoth/sourcemanagement/sourcemanagement.py [hunk 2]
```diff
@@ -41,33 +45,70 @@
...
-                self.service = GitlabService(self.token, instance_url=service_url)
+            if self.service_url:
+                self.service = GitlabService(self.token, instance_url=self.service_url)
             else:
                 self.service = GitlabService(self.token)
             self.repository = self.service.get_project(repo=self.repo, namespace=self.namespace)
         else:
             raise NotImplementedError
 
+    class Token:
```
Comment (fridex) [inline thoth/sourcemanagement/sourcemanagement.py]: Please move this nested class outside - nested classes/functions are hard to test and this introduces unnecessary complexity. You can also use just a function for decoration as you have only one staticmethod.

Comment (saisankargochhayat) [reply thoth/sourcemanagement/sourcemanagement.py]: Thanks, I removed the class.

**Comment by saisankargochhayat** [comment]
@fridex Thanks for the review, any other changes? :smile:

File: thoth/sourcemanagement/github_authentication.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class GithubAuthentication:
+    """Handles getting the OAuth Token for the Github Application."""
+
+    def __init__(self, slug: str) -> None:
+        """Initialize and check values."""
+        self.slug = slug
+
+        self.GITHUB_PRIVATE_KEY_PATH = str(os.getenv("GITHUB_PRIVATE_KEY_PATH"))
```
Comment (fridex) [inline thoth/sourcemanagement/github_authentication.py]: This will make it an instance attribute, can be lowercase based on pep8.

**Comment by fridex** [comment]
> @fridex Thanks for the review, any other changes?

One minor comment, otherwise LGTM.

File: thoth/sourcemanagement/github_authentication.py (hunk 3 above)
Comment (saisankargochhayat) [reply thoth/sourcemanagement/github_authentication.py]: Thanks, I have changed them :)

**Review by fridex — APPROVED** [review]
:+1:

File: thoth/sourcemanagement/github_authentication.py (hunk 3 above)
Comment (fridex) [reply thoth/sourcemanagement/github_authentication.py]: awesome, thanks! :+1:

**Pull request merged**